### PR TITLE
Saas <- Master

### PR DESCRIPTION
**Pull request by ItsSudip**
## Description of the change

Saas <- Master

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/one_signal/transform.js [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+  endpoint = `${endpoint}/${appId}/users/${externalUserId}`;
+  const payload = {};
+  const tags = {};
+  tags.groupId = groupId;
+
+  // Populating tags using allowed properties(from dashboard)
+  const properties = getFieldValueFromMessage(message, "traits");
+  if (properties && allowedProperties && Array.isArray(allowedProperties)) {
+    allowedProperties.forEach(item => {
+      if (
```
Comment (koladilip) [inline v0/destinations/one_signal/transform.js:206]: outer if condition is not required as inner condition is sufficient

File: v0/destinations/one_signal/transform.js [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+const processRouterDest = inputs => {
+  if (!Array.isArray(inputs) || inputs.length <= 0) {
+    const respEvents = getErrorRespEvents(null, 400, "Invalid event array");
+    return [respEvents];
+  }
+
+  const respList = Promise.all(
+    inputs.map(async input => {
+      try {
+        if (input.message.statusCode) {
```
Comment (koladilip) [inline v0/destinations/one_signal/transform.js:261]: let message = input.message.statusCode ? input.message : process(input)
return getSuccessRespEvents(
           message,
           [input.metadata],
           input.destination
 );

File: v0/destinations/canny/transform.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+const processRouterDest = async inputs => {
+  if (!Array.isArray(inputs) || inputs.length <= 0) {
+    const respEvents = getErrorRespEvents(null, 400, "Invalid event array");
+    return [respEvents];
+  }
+
+  const respList = await Promise.all(
+    inputs.map(async input => {
+      try {
+        if (input.message.statusCode) {
```
Comment (koladilip) [inline v0/destinations/canny/transform.js:180]: let message = input.message.statusCode ? input.message : process(input)
return getSuccessRespEvents(
message,
[input.metadata],
input.destination
);

File: v0/destinations/one_signal/transform.js [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+  eg. tags: {
+    "event_name": true
+  }
+  */
+  tags[event] = true;
+  // Populating tags using allowed properties(from dashboard)
+  const properties = get(message, "properties");
+  if (properties && allowedProperties && Array.isArray(allowedProperties)) {
+    allowedProperties.forEach(item => {
+      if (
```
Comment (koladilip) [inline v0/destinations/one_signal/transform.js:156]: if (
          typeof properties[item.propertyName] === "string"
       ) {
        let tagName = eventAsTags ? `${event}_${[item.propertyName]}` : item.propertyName
        tags[tagName] = properties[item.propertyName];
       }
     });

File: v0/destinations/one_signal/transform.js [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  }
+  return response;
+};
+
+const process = event => {
+  return processEvent(event.message, event.destination);
+};
+
+const processRouterDest = inputs => {
+  if (!Array.isArray(inputs) || inputs.length <= 0) {
```
Comment (koladilip) [inline v0/destinations/one_signal/transform.js:253]: inputs.length can be -ve?

File: v0/destinations/canny/util.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      message.context?.traits?.email ||
+      message.properties?.email;
+    const userId = message.userId;
+
+    const header = {
+      "Content-Type": "application/x-www-form-urlencoded",
+      Accept: "application/json"
+    };
+
+    if (email) {
```
[the remark this answers is not in the document]
Comment (utsabc) [reply v0/destinations/canny/util.js:32]: @koladilip we should take it up as separately later, as this is already a release cut. For any code improvements we can take it up in feature review step before it gets merged to master

File: v0/destinations/one_signal/transform.js (hunk 5 above)
Comment (utsabc) [reply v0/destinations/one_signal/transform.js:253]: We never would get this scenario, I think they added this case as a guard block

**Review by koladilip — COMMENTED** [review]
Let's handle these in a separate PR

**Review by utsabc — APPROVED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Pull request merged**